sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -1691,7 +1691,19 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
   override def visitWindowDef(ctx: WindowDefContext): WindowSpecDefinition = withOrigin(ctx) {
     // CLUSTER BY ... | PARTITION BY ... ORDER BY ...
     val partition = ctx.partition.asScala.map(expression)
-    val order = ctx.sortItem.asScala.map(visitSortItem)
+    val order = if (ctx.sortItem.asScala.nonEmpty) {
+      ctx.sortItem.asScala.map(visitSortItem)
+    } else if (ctx.windowFrame != null &&
+      ctx.windowFrame().frameType.getType == SqlBaseParser.RANGE) {
+      // for RANGE window frame, we won't add default order spec
+      ctx.sortItem.asScala.map(visitSortItem)
+    } else {
+      // Same default behaviors like hive, when order spec is null
+      // set partition spec expression as order spec
+      ctx.partition.asScala.map { expr =>
+        SortOrder(expression(expr), Ascending, Ascending.defaultNullOrdering, Set.empty)
```

Comment:
Wait .. why do we set the ordering column as partition column? We should just leave it unspecified so only (non-window) aggregation functions work together with unbounded windows so it doesn't get affected by the order. This is what Scala API does.

Reply:
> Wait .. why do we set the ordering column as partition column? We should just leave it unspecified so only (non-window) aggregation functions work together with unbounded windows so it doesn't get affected by the order. This is what Scala API does.

emmmm， hive doing like this...for me, when user not set order by clause, means he don't care about result order.  For Range DataFrame we can't support this.

Reply:
But the results will be useless. When can it be useful if the order is indeterministic for the functions dependent on the order .. ?

Reply:
> But the results will be useless. When can it be useful if the order is indeterministic for the functions dependent on the order .. ?

In postgre sql , if we don't specify order column, the result is according to partition column 's default sort order.
```
angerszhu=# explain analyze verbose select id, num, lead(id) over (partition by num) from s4;
                                                    QUERY PLAN                                                     
-------------------------------------------------------------------------------------------------------------------
 WindowAgg  (cost=158.51..198.06 rows=2260 width=12) (actual time=0.107..0.122 rows=6 loops=1)
   Output: id, num, lead(id) OVER (?)
   ->  Sort  (cost=158.51..164.16 rows=2260 width=8) (actual time=0.079..0.081 rows=6 loops=1)
         Output: num, id
         Sort Key: s4.num
         Sort Method: quicksort  Memory: 25kB
         ->  Seq Scan on public.s4  (cost=0.00..32.60 rows=2260 width=8) (actual time=0.057..0.061 rows=6 loops=1)
               Output: num, id
 Planning Time: 0.114 ms
 Execution Time: 0.214 ms

angerszhu=# explain analyze verbose select id, num, lead(id) over (partition by num order by id) from s4;
                                                    QUERY PLAN                                                     
-------------------------------------------------------------------------------------------------------------------
 WindowAgg  (cost=158.51..203.71 rows=2260 width=12) (actual time=0.976..1.017 rows=6 loops=1)
   Output: id, num, lead(id) OVER (?)
   ->  Sort  (cost=158.51..164.16 rows=2260 width=8) (actual time=0.067..0.070 rows=6 loops=1)
         Output: id, num
         Sort Key: s4.num, s4.id
         Sort Method: quicksort  Memory: 25kB
         ->  Seq Scan on public.s4  (cost=0.00..32.60 rows=2260 width=8) (actual time=0.042..0.045 rows=6 loops=1)
               Output: id, num
 Planning Time: 0.155 ms
 Execution Time: 1.208 ms
(10 rows)
```

Reply:
I guess because PostgreSQL can keep the natural order. Spark can't keep the natural order. Is PostgreSQL result deterministic?

Reply:
> deterministic

For same sql, result is deterministic.

And we add partition column as order by column by default can keep result deterministic. 



I meet this problem when migration hive sql to spark sql.

Reply:
I think we should not fix it because Spark side at least the results will be non-deterministic. I doubt if this is good to add this support only because of compatibility with other DMBSes when the output is expected to be useless.

Maybe disallowing it might be a better idea than finding another problem later caused by the different and indeterministic data.

Do you maybe know other cases from other distributed DBMSs such as presto?

Reply:
> I think we should not fix it because Spark side at least the results will be non-deterministic. I doubt if this is good to add this support only because of compatibility with other DMBSes when the output is expected to be useless.
> 
> Maybe disallowing it might be a better idea than finding another problem later caused by the different and indeterministic data.
> 
> Do you maybe know other cases from other distributed DBMSs such as presto?

but in my fix,  we add default order spec, the result will be deterministic.
In origin way, this kind sql can't run since it will get non-deterministic result and is rejected by  https://github.com/apache/spark/blob/a28ed86a387b286745b30cd4d90b3d558205a5a7/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L2773-L2776